src/os/bluestore/bluefs_types.h

```diff
@@ -104,19 +164,42 @@ struct bluefs_fnode_t {
     DENC_DUMP_PRE(bluefs_fnode_t);
     _denc_friend(*this, p);
   }
+
   void decode(ceph::buffer::ptr::const_iterator& p) {
-    _denc_friend(*this, p);
+    DENC_START(2, 2, p);
+    denc_varint(ino, p);
+    denc_varint(size, p);
+    denc(mtime, p);
+    denc(type, p);
+    denc(extents, p);
+    if (struct_v >= 2 && type == WAL_V2) {
```

Comment:
This can be handled similarly to my proposal for bluefs_fnode_delta_t above. Just encode/decode all three fields for struct_v >= 2 and assign legacy type in decode if struct_v == 1